### ISSUE TITLE
Definitions depend on universes of unused context variables

### ISSUE THREAD

**Issue by andreaslyn** [coq/coq#10513]
#### Description of the problem

The following
```
Set Universe Polymorphism.
Set Printing Universes.

Section Test.
Context (T:Type).
Definition z := 0.
End Test.

About z.
```
prints `z@{Top.1} : nat`. The value `z` is parametrized over the universe of `T`. This seems counter intuitive when `z` does not depend on `T`. Is this a bug?

**Comment by SkySkimmer** [comment]
I would call it a questionable behaviour, not a bug.

**Comment by andres-erbsen** [comment]
@samuelgruetter is it possible that this is what cause the universe blowup that made us disable universe polymorphism in bedrock2?

**Comment by samuelgruetter** [comment]
Possibly, but hard to tell without two implementations to compare :wink: 
IIRC, the main pain points with universe polymorphism were debugging universe inconsistencies (`Unset Minimization ToSet` did not always help), tactics which should unify universes but don't, and some cases where unifying two terms which only differ by universes takes very long.
Now if terms depend on fewer universes, these problems would likely cause less trouble.

**Comment by spitters** [comment]
@mattam82 Is this intended behavior, or a bug?
@samuelgruetter Could you provide test cases for those problems with tactics?

**Comment by samuelgruetter** [comment]
@spitters you can find commit messages about universe problems we had with `git log --grep=niverse` in [bedrock2](https://github.com/mit-plv/bedrock2/). Some were just endless debugging of universe inconsistencies (which did not always get mentioned in the commit message, and usually also not resulted in a Coq issue because not actually a bug), but we tried to report most issues. The one issue which tipped the balance for me was https://github.com/coq/coq/issues/9979, and especially the fact that it had already been reported [years before](https://github.com/coq/coq/issues/5481) without ever being fixed. Other issues were https://github.com/coq/coq/issues/9813, https://github.com/coq/coq/issues/9507, https://github.com/coq/coq/issues/9892, https://github.com/coq/coq/issues/9268, https://github.com/coq/coq/issues/9125.

**Comment by mattam82** [comment]
It's required. We cannot know if typing the body of z required constraints introduced by (T : Type) unless we look at the whole typing derivation (inefficient). Failure to record these kind of dependencies is the source of the soundness bug we have with template polymorphism right now.

**Comment by mattam82** [comment]
However, we could have a "thinning" lemma (like for "proof using") if you're willing to say from the start that your definition should not be parameterized over `T` and anything depending on it transitively.

**Comment by mattam82** [comment]
I think Pierre-Marie did work on the congruence problem recently. It's not always clear if we should force universe unifications or fail (forcing one here might result in an unexpected error later): in Ltac2 you get to be explicit. In existing ML tactics, I guess we should provide switches to allow forcing (if the tactic is properly handling its map of evars and universes, the changes will be small).

**Comment by SkySkimmer** [comment]
>It's required. We cannot know if typing the body of z required constraints introduced by (T : Type) unless we look at the whole typing derivation (inefficient). Failure to record these kind of dependencies is the source of the soundness bug we have with template polymorphism right now.

We could probably do it if we put the section universes needed by a constant next to the variables it uses and so on, but currently the kernel can't tell apart section universes and regular global universes so it's not easy to do. Maybe when sections are done by the kernel we can come back to it.

**Comment by mattam82** [comment]
Indeed, we could try to be more fine grained on every definition in the section, if the users puts in enough annotations that we can rely on. E.g. if I have A : Type@{i} and B : Type@{j} and C : Type@{k}, `i <= j <= k` in my section and only want to depend on `i` and `k` and `A` and `C`, we need to ensure somehow that we can forget the middle universe `j` here. That's a non-trivial strenghtening lemma there. I'll believe such thinnings are sound when we have proved they're ok in MetaCoq :)

**Comment by spitters** [comment]
Context: This makes combining math-comp and HoTT very difficult.
https://github.com/HoTT/HoTT/issues/1048

**Comment by SkySkimmer** [comment]
That issue doesn't explain where the difficulty appears, can you expand?

**Comment by andreaslyn** [comment]
We are experiencing the universe blowup when building math comp with HoTT. The main cause of the universe blowup is that new polymorphic inductive types are using a lot of universes. For example,
```
prod : Type@{i} -> Type@{j} -> Type@{max(i,j)}
```
uses two universes rather than one. Then
```
Inductive p := p1 : forall A B, prod A B -> p | p2 : forall A B, prod A B -> p
```
uses 4 universes and
```
Inductive q := q1 : p -> q | q2 : p -> q
```
uses 8 universes, etc. This is the main cause of the universe blowup.

The issue with definitions depending on universes of unused context variables is another cause of universe blowup.

**Comment by ppedrot** [comment]
@andreaslyn slightly hijacking this thread, but I have a patch that solves this kind of blowup in a reasonable way. It essentially emulates monomorphism locally to a definition, i.e. every global that has no explicit universe instance will share the same one everywhere in the subterm. I can push it somewhere if you want to try.

**Comment by andreaslyn** [comment]
@ppedrot Yes, that may be useful. I need to build HoTT with that branch also. Is it possible to enable that feature selectively, just for building math components?

**Comment by spitters** [comment]
The problems are in the math-components part. @ppedrot would ltac2 give us anything at this point?
Is it in any way related to the ssr tactics?

**Comment by andreaslyn** [comment]
A workaround for the universe blowup would be to simply unset universe polymorphism globally, but this does not work in places where impredicativity of `Prop` is used, and it also introduces other unexpected universe errors.

I have therefore settled with an approach, which uses universe polymorphism by default and make cartain definitions monomorphic to avoid the universe blowup. This is a daunting task because most definitions are nested inside sections in math components. Mixing monomorphic and polymorphic definitions inside a section is not allowed, so I often have to split up sections in order to make a definition monomorphic. 

I am currently building a significant part of math components and it seems that this approach is going to work out (finally!). The only problem is the huge diff: https://github.com/math-comp/math-comp/compare/master...andreaslyn:hott2, which is, to some extent, caused by splitting up sections.

Is there a way to mix monomorphic and polymorphic definitions in one section?

**Comment by SkySkimmer** [comment]
Where does mathcomp have sections where some definition ends up with an extra universe?

>Is there a way to mix monomorphic and polymorphic definitions in one section?

no

**Comment by SkySkimmer** [comment]
btw you could have a way smaller diff by using Unset Universe Polymorphism instead of annotating each element with Monomorphic.

**Comment by mattam82** [comment]
Instead of that daunting task it might be easier to work on a more fine-grained dependency analysis for polymorphic definitions inside sections. Suppose you introduce a polymorphic universe variable i in a section, what needs to be checked to know if some definition must be discharged w.r.t. it is that it does not appear (recursively) in the bodies, types or constraints of its dependencies in the same section. The kernel will reject you if you thin out a variable that’s necessary for typechecking. In addition, one could have a variant of proof using that will check at the beginning of a proof that some universe from the section can be cleared out, by the same criterion on the statement of the proof it is starting. @ppedrot, since you’re looking at it currently, does that sound right to you?

**Comment by mattam82** [comment]
In addition, regarding other universe blowup sources, I think it’s about time to implement a `Set Polymorphic Universe Cumulativity Explicit` that would by default drop from Cumulativity to Convertibility when encountering constraints involving a (flexible) universe variable by default (when possible, e.g. when the two universes are not algebraic). This would certainly drastically reduce the number of universes and render minimization less important. One could always annotate uses of Cumulativity where needed, and with polymorphic cumulative inductive types it shouldn’t matter much (for universes coming from parameters of inductives which are in many cases irrelevant).

**Comment by ejgallego** [comment]
A small comment that I hope is not too offtopic is about incremental / parallel type-checking.

IMVHO whatever system is used should somehow to be compatible with processing proof bodies in parallel in a sound way. IMHO proof using is quite annoying for users, I guess what you want here is a kind of 2-pass system, where you first process the bodies, then you can [without repeating the type-checking] complete the global certificate for a section.

**Comment by gares** [comment]
Cc @CohenCyril @ggonthier  may be interested to know about this effort on mathcomp and the progress done so far

**Comment by andreaslyn** [comment]
@ppedrot You mentioned a branch which solves the universe blowup in a reasonable way. Is it in a state where I can try it out?

**Comment by ppedrot** [comment]
@andreaslyn There is my publicly available branch https://github.com/ppedrot/coq/tree/poly-emulate-mono-univs but it's lagging behind master quite a bit. I'll rebase it (or do it if you want, it shouldn't be that hard) but in the meantime if you want to test it, set the option `Universe Polymorphism Monomorphism Emulation` to enable the universe sharing.

**Comment by spitters** [comment]
For future reference. Draft of minimization:
https://github.com/coq/coq/pull/11512

**Comment by jdchristensen** [comment]
For the record, while this might be hard to solve in general, if something can be done to handle the case where a universe variable appears nowhere in the definition and has no constraints with any universe varaibles that do appear, that would already be progress.

**Comment by Alizter** [comment]
@mattam82 how does your new unipoly algo address this issue?

**Comment by SkySkimmer** [comment]
It's unrelated
